src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1559,6 +1559,20 @@ LockPartitionsForDistributedPlan(DistributedPlan *distributedPlan)
 		Oid targetRelationId = distributedPlan->targetRelationId;
 
 		LockPartitionsInRelationList(list_make1_oid(targetRelationId), RowExclusiveLock);
+
+		if (PartitionTable(targetRelationId))
+		{
+			Oid parentRelationId = PartitionParentOid(targetRelationId);
+
+			/*
+			 * We lock the parent relation after locking relations to prevent
+			 * distributed deadlock.
+			 * Postgres doesn't take AccessShareLock on the parent table when the
```

Comment:
I think this comment is too technical to follow. Can we tell something like `Postgres only takes when the session accesses the partition first time. That's why we conclude that it is OK to take the lock from PGs perspective. In general case, even though we diverge from Postgres regarding concurrent modifications on partitions vs CREATE/DROP other partition(s), we consider this as a reasonable trade-off for now` (with better sentences)

Reply:
@onderkalaci / @SaitTalhaNisanci - Does this mean that with this change, concurrent updates on same partitions are going to wait for each for other to complete?

Reply:
@hlakshmi no, this change would not change that behaviour.  The change is more around concurrent DROP/CREATE vs UPDATE.

> , concurrent updates on same partitions are going to wait for each for other to complete?

But, the multi-shard updates on the partitions are already serialized (e.g., not allowed concurrently). To try that:

```SQL
BEGIN;
 update test_7_1 set destination = 1000;
UPDATE 0
Time: 103.998 ms
```

```SQL
-- on another session, updating the same partition simply blocks until the above is committed.
update test_7_1 set destination = 1000;
```

There are ways to allow concurrency of these operations, but they are likely to distributed deadlock if touches the same rows. See the setting `set citus.enable_deadlock_prevention to off;`. See the discussion here: https://github.com/citusdata/citus/pull/2494

Reply:
Thanks @onderkalaci